3p/iframe-transport-client.js

```diff
@@ -33,61 +33,134 @@ export class IframeTransportClient {
     /** @private {!Window} */
     this.win_ = win;
 
-    /** @private {?function(string,string)} */
-    this.listener_ = null;
+    /** @private {Object<string,IframeTransportContext>} */
+    this.creativeIdToContext_ = {};
+
+    const parsedFrameName = tryParseJson(this.win_.name);
+
+    /** @private {string} */
+    this.vendor_ = dev().assertString(parsedFrameName['type'],
+        'Parent frame must supply vendor name as type in ' +
+        this.win_.location.href);
+    // Note: amp-ad-exit will validate the vendor name before performing
+    // variable substitution, so if the vendor name is not a valid one from
+    // vendors.js, then its response messages will have no effect.
+    dev().assert(this.vendor_.length, 'Vendor name cannot be empty in ' +
+        this.win_.location.href);
 
     /** @protected {!IframeMessagingClient} */
-    this.client_ = new IframeMessagingClient(win);
-    this.client_.setHostWindow(this.win_.parent);
-    this.client_.setSentinel(dev().assertString(
-        tryParseJson(this.win_.name)['sentinel'],
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setHostWindow(this.win_.parent);
+    this.iframeMessagingClient_.setSentinel(dev().assertString(
+        parsedFrameName['sentinel'],
         'Invalid/missing sentinel on iframe name attribute' + this.win_.name));
-    this.client_.makeRequest(
+    this.iframeMessagingClient_.makeRequest(
         MessageType.SEND_IFRAME_TRANSPORT_EVENTS,
         MessageType.IFRAME_TRANSPORT_EVENTS,
         eventData => {
           const events =
-              /**
-               * @type
-               * {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
-               */
-              (eventData['events']);
-          user().assert(events,
+          /**
+           * @type
+           *   {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
+           */
+          (eventData['events']);
+          dev().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              dev().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              this.contextFor_(event.creativeId).dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+                  e);
```

Comment:
nit: should be a string, use `e.message`

Reply:
Oops, this was in response to @keithwrightbos 's comment at https://github.com/ampproject/amphtml/pull/11306#discussion_r140839676
Removing ".message" was intentional, but I should have also changed the preceding '+' to ','
Fixed.